VDA-XP-Reward/contracts/VDAXPReward.sol

```diff
@@ -0,0 +1,277 @@
+     * @dev Separated for `stack too deep`
+     * @param didAddress - DID address, Used to revert
+     * @param proofSigner - Signer of proof
+     * @param info - Claim information
+     */
+    function _validateProofSigner(address didAddress, address proofSigner, ClaimInfo calldata info) internal virtual view {
+        bool isVerified;
+        uint index;
+        uint length = _trustedSigners.length();
+
+        while (index < length && !isVerified) {
+            if (proofSigner == _trustedSigners.at(index)) {
+                isVerified = true;
+            }
+            unchecked { ++index; }
+        }
+
+        if (!isVerified) {
+            revert InvalidProof(didAddress, info.typeId, info.xp, info.signature);
+        }
+    }
+
+    /**
+     * @notice Check the proofs and return the total amount of XP
+     * @param didAddress - DID address
+     * @param infos - Array of claim requests
+     * @return uint - Total amount of requested claim
+     */
+    function _validateClaimSignature(address didAddress, ClaimInfo[] calldata infos) internal virtual returns(uint) {
+        uint totalXP;
+
+        bytes memory rawMsg;
+        bytes32 dataHash;
+        address proofSigner;
+
+        uint infoLen = infos.length;
+        DateTime._DateTime memory curTime = DateTime.parseTimestamp(block.timestamp);
+
+        for (uint i; i < infoLen;) {
+            if (infos[i].xp == 0) {
+                revert InvalidXP(didAddress, infos[i].typeId, infos[i].xp);
+            }
+            _validateProofTime(didAddress, curTime, infos[i]);
+
+            rawMsg = abi.encodePacked(didAddress, infos[i].typeId, infos[i].issueYear, infos[i].issueMonth);
+            
+            // Check whether `typeId` is claimed in this month
+            if (isClaimedTypeId[rawMsg]) {
+                revert DuplicatedRequest(didAddress, infos[i].typeId, infos[i].xp, infos[i].signature);
+            }
+            isClaimedTypeId[rawMsg] = true;
+
+            if (bytes(infos[i].uniqueId).length != 0) {
+                rawMsg = abi.encodePacked(infos[i].typeId, infos[i].uniqueId);
+                // Check whether `uniqueId` is claimed before
+                if (isClaimedUniqueId[rawMsg]) {
+                    revert DuplicatedUniqueId(didAddress, infos[i].typeId, infos[i].uniqueId, infos[i].xp);
+                }
+                isClaimedUniqueId[rawMsg] = true;
+
+                rawMsg = abi.encodePacked(didAddress, rawMsg, infos[i].issueYear, infos[i].issueMonth, infos[i].xp);
+                
+            } else {
+                rawMsg = abi.encodePacked(rawMsg, infos[i].xp);
+                // Check `Proof` is claimed
+                if (isClaimedSignature[infos[i].signature]) {
+                    revert DuplicatedRequest(didAddress, infos[i].typeId, infos[i].xp, infos[i].signature);
+                }
+                isClaimedSignature[infos[i].signature] = true;
+
+            }
+
+            dataHash = keccak256(rawMsg);
```

Comment:
This should use `VDAVerificationBase.verifyData()`?

Reply:
No, this is not the same as the`verifyData()` function.
```
function verifyData( bytes memory data,  bytes memory signature, bytes memory proof)
```
The `verifyData()` function recovers the `contextSigner` from the `data` and `signature`. And then checks this `${trustedSigner}${strContextSigner}` is valid by the `proof`.

By the way, in the `XPRewardContract`, it check the following:
```
dataHash = keccak256(`${didAddress}${typeId}${uniqueId}${year}${month}${xp}`)
signer = recover(dataHash, claimInfo[i].signature)
```
And checks this `signer` is one of the tursted signers.

Reply:
The code is incorrect then, as the `verifyData()` function should be used.

The `infos[i].signature` will be the signature from a trusted context signer, just like our other reward contract.

Reply:
`infos[i].signautre` is from the trusted signer, correct? Because, it's as following:
```ts
infos[i].signature = TrustedSigner.sign(`${didAddress}${typeId}${uniqueId}${issueYear}${issueMonth}${xp}`)
```
The result of `recover(infos[i].signature)` is the address of the proof signer.
We should check this recoved address is one of the trusted signers.

Reply:
I see the confusion.

You are assuming the truster signer is signing using the `DID` master key.

It should be that the trusted signer is signing using a context signing key. Hence, you should be using the `verifyData()` method which verifies using the context signing key, not the master DID key.

Please ensure the other reward contract is also using the context signing key.